exec/java-exec/src/main/java/org/apache/drill/exec/ops/BaseOperatorContext.java

```diff
@@ -158,25 +159,26 @@ public void close() {
     } catch (RuntimeException e) {
       ex = ex == null ? e : ex;
     }
-    try {
-      if (fs != null) {
+
+    for (DrillFileSystem fs : fileSystems) {
+      try {
         fs.close();
-        fs = null;
-      }
-    } catch (IOException e) {
+      } catch (IOException e) {
       throw UserException.resourceError(e)
-        .addContext("Failed to close the Drill file system for " + getName())
-        .build(logger);
+          .addContext("Failed to close the Drill file system for " + getName())
+          .build(logger);
+      }
     }
+
     if (ex != null) {
       throw ex;
     }
   }
 
   @Override
   public DrillFileSystem newFileSystem(Configuration conf) throws IOException {
-    Preconditions.checkState(fs == null, "Tried to create a second FileSystem. Can only be called once per OperatorContext");
-    fs = new DrillFileSystem(conf, getStats());
+    DrillFileSystem fs = new DrillFileSystem(conf, getStats());
```

Comment:
@parthchandra this is definitely a good question. I did so because in previous code new fs was created for each Hive table split [1]. Projection pusher is used to define fs for each split, it resolves path for table partitions. Frankly saying it worked fine for me without it (all tests have passed) but in Hive code the same approach is used and apparently for the same reasons it was used in Drill. To be safe, I have done the same. If you think we can the same fs for each row group in Hive, then I can adjust the changes.

[1] https://github.com/apache/drill/blob/master/contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/HiveDrillNativeScanBatchCreator.java#L112

Reply:
I'm not suggesting we use the same fs for each split, but the opposite. The fs obect used per split/rowgroup should be different so that we get the right fs wait time for every minor fragment. But this change allows more than one fs object per operator context; which we were explicitly preventing earlier. I'm not sure I understand why you needed to change that.

Reply:
When `AbstractParquetScanBatchCreator.getBatch` method is called, it receives one operator context which is used to allow to create only one file system. It also receives `AbstractParquetRowGroupScan` which contains several row groups. Row groups may belong to different files. For Drill parquet files, we create only one fs and use it for to create readers for each row group. That's why it was fine when operator context allowed to create only one fs. But we needed to adjust it for Hive files. For Hive we need to create fs for each file (since config to each file system is different and created using projection pusher), that's why I had to change operator context to allow more then one file system. I have also introduced `AbstractDrillFileSystemManager` which controls number of file systems created. `ParquetDrillFileSystemManager` creates only one (as was done before). `HiveDrillNativeParquetDrillFileSystemManager` creates fs for each file, so when two row groups belong to the same file, they will get the same fs.

But I agree that for tracking fs (i.e. store.parquet.reader.pagereader.async is set to false) this will create mess in calculations. So I suggest the following fix, for Hive we'll always create non tracking fs, for Drill depending on store.parquet.reader.pagereader.async option. Also I'll add checks in operator context to disallow to create more then one tracking fs and to create tracking fs at all when non-tracking is / are already created.